From 4c06658e988f4ced1a7a574ce082a22ef625df56 Mon Sep 17 00:00:00 2001
Message-ID: <4c06658e988f4ced1a7a574ce082a22ef625df56.1791178681.git.yann.morin@orange.com>
From: "Gary E. Miller" <gem@rellim.com>
Date: Tue, 7 Jul 2026 14:23:56 -0700
Subject: [PATCH] clients/gpsprof.py.in: Quote back ticks in title.

Second try.  Also quote "terminal".

Someone could use the back tick to break out of the string and add
gnuplot commnds.

For issue 404.
Reported by: CuB3y0nd, and Wade Sparks <wsparks@vulncheck.com>
Upstream: https://gitlab.com/gpsd/gpsd/-/commit/4c06658e988f4ced1a7a574ce082a22ef625df56
CVE: CVE-2026-58459
Signed-off-by: Yann E. MORIN <yann.morin@orange.com>

---
 clients/gpsprof.py.in | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/clients/gpsprof.py.in b/clients/gpsprof.py.in
index 202214769..e91367ee3 100644
--- a/clients/gpsprof.py.in
+++ b/clients/gpsprof.py.in
@@ -200,7 +200,7 @@ class plotter(object):
 
         # escape ", `, and \n, for gnuplot, to not break strings
         desc = desc.replace('"', '\\042')
-        desc = desc.replace('`', '\\140')
+        desc = desc.replace("\x60", '\\140')
         desc = desc.replace('\n', '')
 
         return desc
@@ -1271,13 +1271,19 @@ if __name__ == '__main__':
             options.title += '\\n' + options.subtitle
         # escape ",  and`, for gnuplot, to not break strings
         options.title = options.title.replace('"', '\\042')
-        options.title = options.title.replace('"', '\\140')
+        options.title = options.title.replace("\x60", '\\140')
         term_opts = ""
         truecolor_terms = ['png', 'sixelgd', 'wxt']
         if options.terminal in truecolor_terms:
             term_opts = 'truecolor'
-        sys.stdout.write("set terminal %s size 800,950 %s\n"
-                         "set termoption enhanced\n"
+
+        # escape ", `, and \n, for gnuplot, to not break strings
+        options.terminal = options.terminal.replace('"', '\\042')
+        options.terminal = options.terminal.replace("\x60", '\\140')
+        options.terminal = options.terminal.replace('\n', '')
+
+        sys.stdout.write('set terminal "%s" size 800,950 %s\n'
+                         'set termoption enhanced\n'
                          % (options.terminal, term_opts))
         # double quotes on title so \n is parsed by gnuplot
         sys.stdout.write('set title noenhanced "%s\\n\\n"\n' % options.title)
-- 
2.43.0

